### ISSUE TITLE
Video Hangouts

### ISSUE THREAD

**Issue by timsueberkrueb** [tim-sueberkrueb/ubuntu-hangups#32]
This will be possible when Oxide 1.11 is released (https://launchpad.net/oxide).

**Comment by f0o** [comment]
Does this also apply to Audio Hangouts?

**Comment by timsueberkrueb** [comment]
@f0o yes.

**Comment by timsueberkrueb** [comment]
Blocked by this bugs:
- https://bugs.launchpad.net/oxide/+bug/1509256
- https://bugs.launchpad.net/oxide/+bug/1441465

**Comment by bittner** [comment]
Just for curiosity's sake: The camera app, what implementation does it use? Why does the camera app work, and still it's so difficult to make this feature available to the rest of the platform (i.e. integrated in the web browser)?

**Comment by timsueberkrueb** [comment]
@bittner Yeah, good question. The problem is not that I can't use the camera using QML (which should be simple) but I'm waiting for the webbrowser implementation as I basically will just embed the Video Hangouts websites and tweak it a bit (maybe custom, touch-friendly call/chat/etc buttons using qml). But I don't know why they're stuck implementing the web api.

**Comment by timsueberkrueb** [comment]
Update: Oxide now allows video and microphone access and I already tried it, Video Hangouts are now possible! \0/

**Comment by f0o** [comment]
\o/ will try it out in a jiffy

On 4 February 2016 13:00:42 CET, "Tim Süberkrüb" notifications@github.com wrote:

> Update: Oxide now allows video and microphone access and I already
> tried it, Video Hangouts are now possible! \0/
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/tim-sueberkrueb/ubuntu-hangups/issues/32#issuecomment-179790313

## 

Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by timsueberkrueb** [comment]
@tdryer do you know which API calls Hangouts does in order to receive the url to a new video hangout and how can I handle the messages sent when an Hangout comes in? Thanks in advance :)

**Comment by tdryer** [comment]
I took a quick look, and these two APIs seem related to opening a call:

```
https://clients6.google.com/hangouts/v1/hangouts/query
https://clients6.google.com/hangouts/v1/hangouts/resolve
```

I'll try getting them added to hangups when I have time.

For handling events, hangups already exposes the [Hangout call events](http://hangups.readthedocs.org/en/stable/protobuf.html#hangoutevent).

**Comment by tdryer** [comment]
I tried calling `resolve` using hangups and it failed with this error:

```
Daily Limit for Unauthenticated Use Exceeded. Continued use requires signup.
```

I suspect this is an issue with the OAuth access token. The iOS client ID used by hangups may not have access to these APIs.

**Comment by timsueberkrueb** [comment]
@tdryer thanks! Wouldn't it be possible to extract the OAuth token the web version uses?

**Comment by tdryer** [comment]
The web client doesn't use OAuth, but since you have a web view, you may be able to have the user log in through the normal web form and just pass the cookies to hangups.

**Comment by timsueberkrueb** [comment]
@tdryer thanks, I'll try that. The cookies would then have to be refreshed each start though, right? Would make startup probably slower but we'll see.

**Comment by tdryer** [comment]
No, you should be able to reuse the cookies. I'm not sure how long they would be valid for though.
